runtime/src/expected_rent_collection.rs

```diff
@@ -0,0 +1,924 @@
+    Scenario 1:
+      slot 78 is a root
+      slot 79 is skipped/not a root
+      slot 80 is a root (account is rewritten/rent collected here along with accounts from slot 79)
+    Scenario 2:
+      slot 78 is a root
+      slot 79 is skipped/not a root
+      slot 80 is skipped/not a root
+      slot 81 is a root (account is rewritten/rent collected here, along with accounts from slots 79 and 81)
+    This gets us to looking for:
+      the first slot where a slot's pubkey_partition >= pubkey_partition_index - 1.
+
+[footnote2:]
+  Describing partition_index within an epoch:
+  example:
+    slot=0       is epoch=0, partition_index=0
+    slot=1       is epoch=0, partition_index=1
+    slot=431,999 is epoch=0, partition_index=431,999
+    slot=432,000 is epoch=1, partition_index=432,000
+    This is NOT technically accurate because of first_normal_slot, but we'll ignore that.
+    EpochSchedule::get_epoch_and_slot_index(slot) calculates epoch and partition_index from slot.
+
+[footnote3:]
+  when we do a rewrite of account data, only this data changes:
+  1. rent_epoch
+  2. computed hash value (which is a function of (data, lamports, executable, owner, rent_epoch) + slot
+  3. into a new append vec that is associated with the slot#
+
+*/
+
+impl ExpectedRentCollection {
+    #![allow(dead_code)]
+    #[allow(clippy::too_many_arguments)]
+    /// it is possible 0.. rewrites were skipped on this account
+    /// if so, return Some(correct hash as of 'storage_slot')
+    /// if 'loaded_hash' is CORRECT, return None
+    pub fn maybe_rehash_skipped_rewrite(
+        loaded_account: &impl ReadableAccount,
+        loaded_hash: &Hash,
+        pubkey: &Pubkey,
+        storage_slot: Slot,
+        epoch_schedule: &EpochSchedule,
+        rent_collector: &RentCollector,
+        stats: &HashStats,
+        max_slot_in_storages_exclusive: Slot,
+        find_unskipped_slot: impl Fn(Slot) -> Option<Slot>,
+        filler_account_suffix: Option<&Pubkey>,
+    ) -> Option<Hash> {
+        let expected = match ExpectedRentCollection::new(
+            pubkey,
+            loaded_account,
+            storage_slot,
+            epoch_schedule,
+            rent_collector,
+            max_slot_in_storages_exclusive,
+            find_unskipped_slot,
+            filler_account_suffix,
+        ) {
+            None => {
+                // use the previously calculated hash
+                return None;
+            }
+            Some(expected) => expected,
+        };
+
+        let recalc_hash = AccountsDb::hash_account_with_rent_epoch(
+            expected.expected_rent_collection_slot_max_epoch,
+            loaded_account,
+            pubkey,
+            expected.rent_epoch,
+        );
+        if &recalc_hash == loaded_hash {
+            // unnecessary calculation occurred
+            stats.rehash_unnecessary.fetch_add(1, Ordering::Relaxed);
+            return None;
+        }
+        stats.rehash.fetch_add(1, Ordering::Relaxed);
```

Comment:
should we move this line before 252? does stats.rehash measures the total number of reshashes?

Reply:
renamed